### PR TITLE
Refs #23130 -- Allowed BooleanField to be null=True

### PR DESCRIPTION
**Pull request by lynncyrin**
https://code.djangoproject.com/ticket/23130

Picks up were @timgraham left off in https://github.com/django/django/pull/8467

refs:
- https://github.com/django/django/pull/3025
- https://code.djangoproject.com/ticket/22282
- https://code.djangoproject.com/ticket/26168
- https://code.djangoproject.com/ticket/22282

details:
- `BooleanField` model field can take `null=True`
- ~`NullBooleanField` is now just `BooleanField(null=True)`~
- depreciates `NullBooleanField` (this decision was made in @timgraham's PR)
- ~normalizes boolean type casting, and incorporates `distutils.util.strtobool`~
- ~collects all the different boolean casting special cases into a single file `django/utils/typecasting.py`~

still TODO:
- [x] there's at least a dozen tests failing <<< resolved
- [x] determine if this is a large enough PR to get added to [AUTHORS](https://github.com/django/django/blob/master/AUTHORS) ^_^ <<< call: no. the PR is much simpler without the `typecasting.py` change
- [ ] check for `NullBooleanField` isinstance calls to turn into `null=True`
- ~`typecasting.py` entrypoint functions need better names~
- ~evaluate `typecasting.py` `nullable=[ True | False ]` handling~
- ~add a few more cases to `test_typecasting.py`~

### CODE REVIEW

**Comment by timgraham** [comment]
To answer your question from IRC: yes, the relocating of logic to `typecasting.py` seems out of scope for this ticket. We prefer to do refactorings in separate pull requests that don't change behavior.
The remaining issues where I left off was:
* As needed, continue adding tests for `BooleanField(null=True)` where `NullBooleanField` is tested.
* Solve the issue with Oracle check constraints. I think the appropriate check constraint for `BooleanField(null=True)` won't be generated because `Field.get_internal_type()` returns `BooleanField` rather than `NullBooleanField`. A possible fix is:
```
def get_internal_type(self):
    return 'NullBooleanField' if self.null else 'BooleanField'
```
However, the fact that there aren't any test failures without that makes me wonder if the constraints are working properly in the first place. \cc @felixxm.

File: django/forms/fields.py [hunk 1]
```diff
@@ -686,22 +686,27 @@ def split_url(url):
 
 
 class BooleanField(Field):
+    """
+    A Field that displays a Checkbox by default, and
+    returns True or False
+
+    But if given null=False, it displays a Select instead and
```
Comment (jonashaag) [inline django/forms/fields.py]: Typo: `null=True`

**Comment by felixxm** [comment]
@timgraham `NULL` cannot be compare with standard operators on Oracle, it works only with `IS NULL` and `IS NOT NULL` operators. Therefore constraint `'%(qn_column)s IN (0,1)'` works as: if column value is not null then it must be one of values `(0,1)` so it's basically the same as `'(%(qn_column)s IN (0,1)) OR (%(qn_column)s IS NULL)'`. We check nullability with separate `NOT NULL` constraint. I've proposed to unify these checks on Oracle,  #9018.

**Comment by lynncyrin** [comment]
Went through the last set of review comments and shrunk the PR's scope

Still TODO: failing tests

**Comment by lynncyrin** [comment]
✔️  tests passing

> Solve the issue with Oracle check constraints. I think the appropriate check constraint for BooleanField(null=True) won't be generated because Field.get_internal_type() returns BooleanField rather than NullBooleanField. A possible fix is...

Implemented that fix! But also I feel this becomes less relevant if https://github.com/django/django/pull/9018 is merged first

> As needed, continue adding tests for BooleanField(null=True) where NullBooleanField is tested.

So ^ is the last thing to do

Although I'm not sure to what extent adding those tests would be overkill, since `NullBooleanField` is now just

https://github.com/lynncyrin/django/blob/254d3149cb50016ae6e0746488b1042f5a05d8ec/django/forms/fields.py#L742-L748

**Comment by timgraham** [comment]
1. Yes, I think we can omit the change to `get_internal_type()` considering https://github.com/django/django/pull/9018.
2. I don't think the form field changes to put `NullBooleanField`'s logic in `BooleanField` is much of a simplification. Let's focus on the model field and leave the form fields alone for now.
3. The idea with the additional tests is about the model field. For example, in `tests/bulk_create/models.py` there's a usage of `NullBooleanField`. A similar test for `BooleanField(null=True)` should be added there.

**Comment by lynncyrin** [comment]
> omit the change to get_internal_type() considering #9018

✔️ 

> focus on the model field and leave the form fields alone for now

✔️ 

> additional tests [for] the model field

WIP. I added the `bulk_create` one (it was obvious what to do there). I still need to appraise what adequate tests for be for: 

- postgres_tests
- test_promises
- test_booleanfield
- managers_regress
- generic_relations_regress
- expressions_case
- annotations

**Comment by lynncyrin** [comment]
In case I forget to make them, some optimization ticket ideas spawned from this PR are:

- make all boolean type casting utilize `distutils.util.strtobool`, and collect form field and model field special cases into a single boolean type casting file
- make `NullBoolean` model field `BooleanField(null=True)`
- make `NullBoolean` form field `BooleanField(null=True)`

**Comment by timgraham** [comment]
I incorporated some of these updates in https://github.com/django/django/pull/8467 which should be ready for review now. Thanks!

**Pull request closed without merging**